File: rerun-reports/lane-a/034/20260828-035552/results/storage/data-plane/java/crud/storage-dp-java-blob-manager/java-azure-skills/azure-skill-mcp/generated-code/src/main/java/com/example/blobmanager/BlobStorageService.java
package com.example.blobmanager;

import com.azure.storage.blob.BlobClient;
import com.azure.storage.blob.BlobContainerClient;
import com.azure.storage.blob.BlobServiceClient;
import com.azure.storage.blob.models.BlobItem;
import com.azure.storage.blob.models.BlobProperties;
import com.azure.storage.blob.options.BlobParallelUploadOptions;
import com.azure.storage.blob.specialized.BlobLeaseClient;
import com.azure.storage.blob.specialized.BlobLeaseClientBuilder;

import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public final class BlobStorageService {
    private final BlobContainerClient containerClient;

    public BlobStorageService(BlobServiceClient serviceClient, String containerName) {
        Objects.requireNonNull(serviceClient, "serviceClient");
        this.containerClient = serviceClient.getBlobContainerClient(
                Objects.requireNonNull(containerName, "containerName"));
    }

    public void ensureContainerExists() {
        containerClient.createIfNotExists();
    }

    public BlobProperties upload(
            String blobName,
            Path source,
            Map<String, String> metadata,
            Map<String, String> tags
    ) {
        return upload(blobName, source, metadata, tags, null, null);
    }

    public BlobProperties upload(
            String blobName,
            Path source,
            Map<String, String> metadata,
            Map<String, String> tags,
            String expectedETag,
            String leaseId
    ) {
        BlobClient blobClient = containerClient.getBlobClient(blobName);
        BlobParallelUploadOptions options = BlobUploadOptionsFactory.create(
                source, metadata, tags, expectedETag, leaseId);

        blobClient.uploadWithResponse(options, null, null);
        return blobClient.getProperties();
    }

    public void download(String blobName, Path destination, boolean overwrite) {
        containerClient.getBlobClient(blobName).downloadToFile(destination.toString(), overwrite);
    }

    public List<BlobItem> listBlobs() {
        return containerClient.listBlobs().stream().toList();
    }

    public boolean delete(String blobName) {
        return containerClient.getBlobClient(blobName).deleteIfExists();
    }

    public Lease acquireLease(String blobName, int durationSeconds) {
        BlobLeaseClient leaseClient = new BlobLeaseClientBuilder()
                .blobClient(containerClient.getBlobClient(blobName))
                .buildClient();
        return new Lease(leaseClient.acquireLease(durationSeconds), leaseClient);
    }

    public static final class Lease implements AutoCloseable {
        private final String leaseId;
        private final BlobLeaseClient leaseClient;
        private boolean released;

        private Lease(String leaseId, BlobLeaseClient leaseClient) {
            this.leaseId = leaseId;
            this.leaseClient = leaseClient;
        }

        public String leaseId() {
            return leaseId;
        }

        @Override
        public void close() {
            if (!released) {
                leaseClient.releaseLease();
                released = true;
            }
        }
    }
}

File: rerun-reports/04-js-ts/20260829-141457/results/storage/data-plane/js-ts/crud/storage-dp-js-ts-blob-manager/js-ts-azure-skills/azure-skill-mcp-microsoft-skill/generated-code/src/blob-storage-service.ts
import { createReadStream, createWriteStream } from "node:fs";
import { mkdir } from "node:fs/promises";
import { dirname } from "node:path";
import { pipeline } from "node:stream/promises";
import {
  type BlobItem,
  type BlobServiceClient,
  type BlockBlobClient,
  RestError,
} from "@azure/storage-blob";

export interface UploadOptions {
  metadata?: Record<string, string>;
  tags?: Record<string, string>;
  contentType?: string;
  onProgress?: (uploadedBytes: number) => void;
}

export interface BlobStorageServiceOptions {
  uploadBufferSize?: number;
  uploadConcurrency?: number;
  leaseWaitMs?: number;
  leasePollMs?: number;
}

const DEFAULT_UPLOAD_BUFFER_SIZE = 8 * 1024 * 1024;
const DEFAULT_UPLOAD_CONCURRENCY = 5;
const DEFAULT_LEASE_WAIT_MS = 30_000;
const DEFAULT_LEASE_POLL_MS = 1_000;

export class BlobStorageService {
  private readonly containerClient;
  private readonly uploadBufferSize: number;
  private readonly uploadConcurrency: number;
  private readonly leaseWaitMs: number;
  private readonly leasePollMs: number;

  public constructor(
    blobServiceClient: BlobServiceClient,
    containerName: string,
    options: BlobStorageServiceOptions = {},
  ) {
    this.containerClient = blobServiceClient.getContainerClient(containerName);
    this.uploadBufferSize =
      options.uploadBufferSize ?? DEFAULT_UPLOAD_BUFFER_SIZE;
    this.uploadConcurrency =
      options.uploadConcurrency ?? DEFAULT_UPLOAD_CONCURRENCY;
    this.leaseWaitMs = options.leaseWaitMs ?? DEFAULT_LEASE_WAIT_MS;
    this.leasePollMs = options.leasePollMs ?? DEFAULT_LEASE_POLL_MS;
  }

  public async ensureContainer(): Promise<void> {
    await this.containerClient.createIfNotExists();
  }

  public async upload(
    blobName: string,
    filePath: string,
    options: UploadOptions = {},
  ): Promise<void> {
    const blobClient = this.containerClient.getBlockBlobClient(blobName);

    // An atomic create handles a missing blob; if another writer wins that
    // race, retry through the lease-protected existing-blob path.
    for (;;) {
      if (!(await this.blobExists(blobClient))) {
        try {
          await this.uploadStream(blobClient, filePath, options, undefined, "*");
          return;
        } catch (error: unknown) {
          if (!this.isCreateRace(error)) {
            throw error;
          }
        }
      }

      const leaseClient = blobClient.getBlobLeaseClient();
      await this.acquireLease(leaseClient);
      let operationError: unknown;
      try {
        await this.uploadStream(
          blobClient,
          filePath,
          options,
          leaseClient.leaseId,
        );
        return;
      } catch (error: unknown) {
        operationError = error;
        throw error;
      } finally {
        try {
          await leaseClient.releaseLease();
        } catch (releaseError: unknown) {
          if (operationError === undefined) {
            throw releaseError;
          }
          console.error("Failed to release the blob lease:", releaseError);
        }
      }
    }
  }

  public async download(blobName: string, destinationPath: string): Promise<void> {
    const blobClient = this.containerClient.getBlobClient(blobName);
    const response = await blobClient.download();
    if (!response.readableStreamBody) {
      throw new Error(`Blob ${blobName} did not return a readable stream.`);
    }

    await mkdir(dirname(destinationPath), { recursive: true });
    await pipeline(
      response.readableStreamBody,
      createWriteStream(destinationPath),
    );
  }

  public async list(): Promise<BlobItem[]> {
    const blobs: BlobItem[] = [];
    for await (const blob of this.containerClient.listBlobsFlat({
      includeMetadata: true,
      includeTags: true,
    })) {
      blobs.push(blob);
    }
    return blobs;
  }

  public async delete(blobName: string): Promise<boolean> {
    const response = await this.containerClient
      .getBlobClient(blobName)
      .deleteIfExists({ deleteSnapshots: "include" });
    return response.succeeded;
  }

  private async uploadStream(
    blobClient: BlockBlobClient,
    filePath: string,
    options: UploadOptions,
    leaseId?: string,
    ifNoneMatch?: string,
  ): Promise<void> {
    const source = createReadStream(filePath);
    await blobClient.uploadStream(
      source,
      this.uploadBufferSize,
      this.uploadConcurrency,
      {
        metadata: options.metadata,
        tags: options.tags,
        blobHTTPHeaders: options.contentType
          ? { blobContentType: options.contentType }
          : undefined,
        conditions: {
          leaseId,
          ifNoneMatch,
        },
        onProgress: options.onProgress
          ? ({ loadedBytes }) => options.onProgress?.(loadedBytes)
          : undefined,
      },
    );
  }

  private async blobExists(blobClient: BlockBlobClient): Promise<boolean> {
    try {
      await blobClient.getProperties();
      return true;
    } catch (error: unknown) {
      if (error instanceof RestError && error.statusCode === 404) {
        return false;
      }
      throw error;
    }
  }

  private async acquireLease(
    leaseClient: ReturnType<BlockBlobClient["getBlobLeaseClient"]>,
  ): Promise<void> {
    const deadline = Date.now() + this.leaseWaitMs;
    for (;;) {
      try {
        await leaseClient.acquireLease(-1);
        return;
      } catch (error: unknown) {
        if (!this.isLeaseConflict(error) || Date.now() >= deadline) {
          throw error;
        }
        await new Promise((resolve) => setTimeout(resolve, this.leasePollMs));
      }
    }
  }

  private isCreateRace(error: unknown): boolean {
    return (
      error instanceof RestError &&
      (error.statusCode === 409 || error.statusCode === 412)
    );
  }

  private isLeaseConflict(error: unknown): boolean {
    return (
      error instanceof RestError &&
      error.statusCode === 409 &&
      (error.code === "LeaseAlreadyPresent" ||
        error.code === "LeaseIsBreakingAndCannotBeAcquired")
    );
  }
}
